python/ray/data/iterator.py

```diff
@@ -344,35 +360,63 @@ def iter_torch_batches(
+
+            Args:
+                batch: Input batch to move to device. Can be:
+                    - Dictionary mapping column names to lists of tensors
+                    - Any other type supported by move_tensors_to_device
+
+            Returns:
+                Batch with tensors moved to the target device. Type matches input type:
+                - If input is Dict[str, List[torch.Tensor]],
+                  returns Dict[str, torch.Tensor]
+                - Otherwise returns the same type as input with tensors moved to device
+            """
+            from ray.air._internal.torch_utils import move_tensors_to_device
+
+            return move_tensors_to_device(batch, device=device)
+
+        finalize_fn = None
         if collate_fn is None:
             # The default collate_fn handles formatting and Tensor creation.
-            # Here, we set device=None to defer host to device data transfer
-            # to the subsequent finalize_fn.
-            def collate_fn(batch: Union[np.ndarray, Dict[str, np.ndarray]]):
-                return convert_ndarray_batch_to_torch_tensor_batch(
-                    batch,
-                    dtypes=dtypes,
-                    device=None,
-                )
-
-            # The default finalize_fn handles the host to device data transfer.
-            # This is executed in a 1-thread pool separately from collate_fn
-            # to allow independent parallelism of these steps.
-            def finalize_fn(batch: Union["torch.Tensor", Dict[str, "torch.Tensor"]]):
-                if device is not None:
-                    if isinstance(batch, dict):
-                        for k, t in batch.items():
-                            batch[k] = t.to(device=device)
-                    else:
-                        batch = batch.to(device=device)
-                return batch
-
+            # Here, we defer host to device data transfer to the subsequent
+            # finalize_fn.
+            collate_fn = DefaultCollateFn(
+                dtypes=dtypes,
+                device=device,
+            )
+            finalize_fn = default_finalize_fn
+            batch_format = "pyarrow"
+        elif isinstance(collate_fn, ArrowBatchCollateFn):
+            # The ArrowBatchCollateFn handles formatting and Tensor creation.
+            # Here, we defer host to device data transfer to the subsequent
+            # finalize_fn.
+            finalize_fn = default_finalize_fn
+            batch_format = "pyarrow"
+        elif isinstance(collate_fn, NumpyBatchCollateFn):
+            batch_format = "numpy"
+        elif isinstance(collate_fn, PandasBatchCollateFn):
+            batch_format = "pandas"
+        elif callable(collate_fn):
+            batch_format = "numpy"
+            warnings.warn(
+                "Passing a raw callable `collate_fn` to `iter_torch_batches` is deprecated in Ray 2.46.",
+                DeprecationWarning,
```

Comment:
I think regular `DeprecationWarning` doesn't show up to people easily unless they run with a special flag.

```suggestion
                "Passing a function to `iter_torch_batches(collate_fn)` will be deprecated in Ray 2.46. Please switch to using a callable class that inherits from X, Y, or Z.`",
                RayDeprecationWarning,
```

Reply:
let's follow the deprecation timeline here: https://docs.ray.io/en/latest/ray-contribute/api-policy.html#id3

(1 release of soft deprecation might be too short)